Fix empty edit fields not speaking in Microsoft Edge

LeonarddeR opened the pull request:
### Link to issue number:
Fixes #12474 
Fixes #12189

### Summary of the issue:
Empty edit fields aren't spoken in Edge. There's a bug in the UIA implementation that causes expand to line to fail when there is no text content in the object

### Description of how this pull request fixes the issue:
1. Use the ChromiumUIA object overlay for text boxes in the Chrome/Edge interface. For Edge, this applies by default. For Chrome, this only applies to cases where experimental UIA support is enabled, but this is far from stable in Chrome itself.
2. On the ChromeUIATextInfo when expanding to line, check whether there's any text in the document range. If not, don't try to expand as it will break the range.
3. Make sure not to cache the TextInfo property on UIA objects.ChromiumUIATextInfo. This is necessary because the property is fetched before the overlay classes are added that might change the result of evaluating the property. This ensures the proper TextInfo is used on focus.

### Testing strategy:
Tested that the address bar of Edge is read correctly when empty and focused.

### Known issues with pull request:
None known.

### Change log entries:
Bug fixes
* In the new Chromium based Microsoft Edge, text fields such as the address bar are now announced when empty. (#12474)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

LeonarddeR committed:
Ensure that edit boxes in the Chrome/Edge UIA implementation don't fail when there is empty content.

LeonarddeR committed:
Ensure that the NVDAObject property cache is cleared when overlay classes are added on top of the object

leonardder commented:
I actually think that this pr will mean a significant improvement for Edge users, therefore I'd like to request to have this merged into beta instead.

cary-rowen commented:
This is a significant improvement and hope to be merged into Beta

alexarnaud commented:
Thanks a lot @leonardder for fixing this.

I have two questions : 
* Do you still reproduce with Edge version 91 released today?
* Is it possible for you to report the UIA issue against Edge?

Best regards.

leonardder commented:
@alexarnaud  wrote:

> I have two questions :
> 
> * Do you still reproduce with Edge version 91 released today?

Yes.

> * Is it possible for you to report the UIA issue against Edge?

If I knew where to do that, I'd certainly would. DO I need to do it against Edge or Chromium? Thoughts @michaelDCurran?

josephsl commented:
Hi, I recommend contacting Microsoft Edge team (hopefully through Feedback hub if possible) and also ask that Chromium/Chrome team look into this issue (because this could also impact other Chromium based browsers). Thanks.

leonardder commented:
I think it would be better if NV Access pokes their Microsoft contacts about this.

josephsl commented:
Hi, and Google, too.

leonardder commented:
Well, as long as Google's implementation is experimental, I guess it's up to Microsoft to push fixes upstream.

Brian1Gaff commented:
If one could trace it directly to the Chrome team, surely then it would be 
on all and fixed with the same patch otherwise you might end up with a 
different work around for every browser if one is not careful.
 Brian

***@***.***
Sent via blueyonder.
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
----- Original Message ----- 
From: "Joseph Lee" ***@***.***>
To: "nvaccess/nvda" ***@***.***>
Cc: "Subscribed" ***@***.***>
Sent: Friday, May 28, 2021 2:30 PM
Subject: Re: [nvaccess/nvda] Fix empty edit fields not speaking in Microsoft 
Edge (#12475)


> Hi, and Google, too.
>
>
>
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/12475#issuecomment-850420209

michaelDCurran reviewed (changes requested):
I am concirned that this may introduce performance issues, Especially also as these days it is important to try and keep the number of cross-process calls in UIA to a minimum with scenarios such as WDAG and RAIL.

In ChromiumUIATextInfo.expand: instead of the added calls to clone and compare, why not just call GetText(1) to fetch the first 1 characters and check if it is the empty string or not. If it is empty, then simply don't call expandToEnclosingUnit.
 
Although in general you are probably correct that we should call obj.invalidateCache after applying overlay classes, I would prefer this to be in its own pr, where we can very carefully consider and track any performance issues this may cause. This change obviously affects code across NVDA, not just the Edge address bar.
However, to address the specific issue of TextInfos incorrectly being cached here, perhaps we can at least set `_cache_TextInfo = False` on the UIA NVDAObject which will stop the TextInfo property from being cached. All the major calls it makes internally are cached properties themselves, so this would have no real performance impact, but it would still allow the current value of `_TextInfo` to be used, thus an overlay with a different TextInfo class would correctly be reflected in speakObject after overlay classes were applied but still in the same core cycle.

LeonarddeR committed:
Revert "Ensure that the NVDAObject property cache is cleared when overlay classes are added on top of the object" and set cache attribute on UIA

This reverts commit 800abad405d7713680c9fd329d680bdd083ef6ae.

leonardder commented:
> In ChromiumUIATextInfo.expand: instead of the added calls to clone and compare, why not just call GetText(1) to fetch the first 1 characters and check if it is the empty string or not. If it is empty, then simply don't call expandToEnclosingUnit.

Is that going to work? I assume GetText returns text in the range and doesn't return anything when the range is collapsed, therefore if the range is collapsed, it will never expand.

michaelDCurran commented:
Ah, sorry, I mean self.obj.UIATextPattern.documentRange.GetText(1)

LeonarddeR committed:
Review action

leonardder commented:
Ah, yeah that makes sence. I changed this as requested.
I also tested every UIA unit and it turns out that it only fails when expanding to line as I originally thought. Therefore I changed the code to only apply when expanding to line. I will change the description accordingly.

michaelDCurran committed:
Update what's new
